src/state/token_index.ml

```diff
@@ -0,0 +1,81 @@
+open Yojson
+open Lwt
+open Learnocaml_data
+
+
+let token_file = "token.json"
+
+(* Unlocked *)
+let mutex_token = Lwt_mutex.create ()
+
+let cast_list l = `List l
+
+let cast_string (value:string) = `String value
+
+let string_to_token l = List.map Token.parse l
+
+let get (sync_dir : string) () =
+      let base = sync_dir in
+      let ( / ) dir f = if dir = "" then f else Filename.concat dir f in
+      let rec scan f d acc =
+        let rec aux s acc =
+          Lwt.catch (fun () ->
+              Lwt_stream.get s >>= function
+              | Some ("." | "..") -> aux s acc
+              | Some x -> scan f (d / x) acc >>= aux s
+              | None -> Lwt.return acc)
+          @@ function
+          | Unix.Unix_error (Unix.ENOTDIR, _, _) -> f d acc
+          | Unix.Unix_error _ -> Lwt.return acc
+          | e -> Lwt.fail e
+        in
+        aux (Lwt_unix.files_of_directory (base / d)) acc
+      in
+      scan (fun d acc ->
+          let d =
+            if Filename.basename d = "save.json" then Filename.dirname d
+            else d
+          in
+          let stok = String.map (function '/' | '\\' -> '-' | c -> c) d in
+          if Token.check stok then
+            Lwt.return (stok :: acc)
+          else
+            Lwt.return acc
+        ) "" []
+
+
+let write_file file mutex data =
+  Lwt_mutex.lock mutex >|= fun () ->
+  let oo = open_out file in
+  Yojson.Basic.pretty_to_channel oo data;
+  close_out oo;
+  Lwt_mutex.unlock mutex
+
+let create_index (sync_dir : string) =
+  let l = get sync_dir () in
+  let data =  l >|= List.map cast_string >|= cast_list in
+  data >>= write_file (sync_dir ^ "/" ^ token_file) mutex_token
```

Comment:
it seems the four occurrences of
`sync_dir ^ "/" ^ token_file`
are not portable (under Windows, the directory separator is a `\`)
but one could replace this with `sync_dir / token_file`
by taking advantage of the definition above
`let ( / ) dir f = if dir = "" then f else Filename.concat dir f in`
(which could be promoted to a file-local definition, instead of a function-local definition)

@Aleridia I can push a commit applying this change, unless you prefer to do that yourself.

Reply:
I will be thankfull if you can push it.

I can do it but it won't be before tomorrow.